torchaudio/csrc/sox/utils.cpp

```diff
@@ -188,63 +198,155 @@ torch::Tensor convert_to_tensor(
-      return SOX_ENCODING_UNSIGNED;
-    if (dtype == torch::kInt16)
-      return SOX_ENCODING_SIGN2;
-    if (dtype == torch::kInt32)
-      return SOX_ENCODING_SIGN2;
-    if (dtype == torch::kFloat32)
-      return SOX_ENCODING_FLOAT;
-    throw std::runtime_error("Unsupported dtype.");
+namespace {
+
+std::tuple<sox_encoding_t, unsigned> get_save_encoding_for_wav(
+    const std::string format,
+    const c10::optional<std::string>& encoding,
+    const c10::optional<int64_t>& bits_per_sample) {
+  if (!encoding.has_value()) {
+    if (!bits_per_sample.has_value())
+      return std::make_tuple<>(SOX_ENCODING_SIGN2, 16);
+    auto val = static_cast<unsigned>(bits_per_sample.value());
+    if (val == 8)
+      return std::make_tuple<>(SOX_ENCODING_UNSIGNED, 8);
+    return std::make_tuple<>(SOX_ENCODING_SIGN2, val);
   }
-  if (filetype == "sph")
-    return SOX_ENCODING_SIGN2;
-  if (filetype == "amr-nb")
-    return SOX_ENCODING_AMR_NB;
-  throw std::runtime_error("Unsupported file type: " + filetype);
+  if (encoding == ENCODING_PCM_SIGNED) {
+    if (!bits_per_sample.has_value())
+      return std::make_tuple<>(SOX_ENCODING_SIGN2, 16);
+    auto val = static_cast<unsigned>(bits_per_sample.value());
+    if (val == 8) {
+      TORCH_WARN_ONCE(
+          "%s does not support 8-bit signed PCM encoding. Using 16-bit.",
+          format);
+      val = 16;
+    }
+    return std::make_tuple<>(SOX_ENCODING_SIGN2, val);
+  }
+  if (encoding == ENCODING_PCM_UNSIGNED) {
+    if (!bits_per_sample.has_value())
+      return std::make_tuple<>(SOX_ENCODING_UNSIGNED, 8);
+    auto val = static_cast<unsigned>(bits_per_sample.value());
+    if (val != 8)
+      TORCH_WARN_ONCE(
+          "%s only supports 8-bit for unsigned PCM encoding. Using 8-bit.",
+          format);
+    return std::make_tuple<>(SOX_ENCODING_UNSIGNED, 8);
+  }
+  if (encoding == ENCODING_PCM_FLOAT) {
+    auto val = static_cast<unsigned>(bits_per_sample.value_or(32));
+    if (val != 32)
+      TORCH_WARN_ONCE(
+          "%s only supports 32-bit for floating point PCM encoding. Using 32-bit.",
+          format);
+    return std::make_tuple<>(SOX_ENCODING_FLOAT, 32);
+  }
+  if (encoding == ENCODING_ULAW) {
+    auto val = static_cast<unsigned>(bits_per_sample.value_or(8));
+    if (val != 8)
+      TORCH_WARN_ONCE(
+          "%s only supports 8-bit for mu-law encoding. Using 8-bit.", format);
+    return std::make_tuple<>(SOX_ENCODING_ULAW, 8);
+  }
+  if (encoding == ENCODING_ALAW) {
+    auto val = static_cast<unsigned>(bits_per_sample.value_or(8));
+    if (val != 8)
+      TORCH_WARN_ONCE(
+          "%s only supports 8-bit for a-law encoding. Using 8-bit.", format);
+    return std::make_tuple<>(SOX_ENCODING_ALAW, 8);
+  }
+  std::ostringstream message;
+  message << format
+          << " format does not support encoding: " << encoding.value();
+  throw std::runtime_error(message.str());
+}
+
+std::tuple<sox_encoding_t, unsigned> get_save_encoding(
```

Comment:
This function implements the mapping from user-provided encoding configuration to the actual encoding configuration.